blog/_src/posts/2016-07-12-tutorial-zero-to-sixty-in-racket.scrbl

```diff
@@ -0,0 +1,543 @@
+        (circular-shift '("A" "B" "C"))
+        '("B" "C" "A")))
+
+    (define (alphabetize all-shifts)
+      (sort all-shifts shift<?))
+
+    (module+ test
+      (check-equal?
+        (alphabetize '(("racket" "is")
+                       ("as")
+                       ("racket" "does")))
+        '(("as")
+          ("racket" "does")
+          ("racket" "is"))))
+))
+
+
+@section{Recognizing Patterns, Avoiding Repetition}
+
+Every unit test needs to go through @racket[check-equal?].
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+
+    (module+ test
+      (check-equal?
+        (kwik-split '())
+        '())
+      (check-equal?
+        (kwik-split '("hello    world"))
+        '(("hello" "world")))
+      (check-equal?
+        (kwik-split '(" lost " " in " "space"))
+        '(("lost") ("in") ("space")))
+      (check-equal?
+        (kwik-split '("something"))
+        '()))
+))
+
+These tests follow a simple pattern that we can express as a @emph{syntax rule}.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+
+    (module+ test
+      (define-syntax-rule (check-equal?* [i o] ...)
+        (begin
+          (check-equal? i o)
+          ...))
+
+      (check-equal?*
+        [(kwik-split '())
+         '()]
+        [(kwik-split '("hello    world"))
+         '(("hello" "world"))]
+        [(kwik-split '(" out " " in " "the ozone"))
+         '(("out") ("in") ("the" "ozone"))]
+        [(kwik-split '("something"))
+         '()]))
+))
+
+Those @racket[...] are not pseudocode!
+They denote Kleene-star repetition, like a sextile (@racket{*}) in a regular expression.
+In this case, the input pattern is just a list of two S-expressions, @racket[i] and
+ @racket[o].
+Uses of @racket[i] and @racket[o] must be followed by one @racket[...] to splice
+ the captured S-expressions into the result.
+
+Many languages offer higher-order functions and polymorphism to abstract common
+ behaviors.
+Syntax extensions are a different, useful way to avoid repeating yourself.
+
+
+@section{Adding Static Types}
+
+Changing the @hash-lang[] line to Typed Racket adds static type-checking to
+ our program.
+If we only change the language and run the code as-is, there will be type errors.
+But we can use submodules again to incrementally arm our design with types.
+
+@(begin
+  #reader scribble/comment-reader
+  (racketblock
+    lang racket
+
+    (module t typed/racket
+      ; Need to annotate:
+      ; - function parameters
+      ; - for-loop return types
+
+      (: kwik-read : Path-String -> (Listof String))
+      (define (kwik-read filename)
+        (with-input-from-file filename
+          (λ ()
+            (for/list ([line (in-lines)])
+                      : (Listof String)
+              line))))
+
+      ; For later: move more untyped functions here
+
+      (provide (all-defined-out)))
+    (require 't)
+
+    (define (kwik-split lines)
+      (map string-split lines))
+
+    ; <rest of file omitted>
```

Comment:
I think it may be worth being a bit more explicit here, giving the other functions (it's not so much code) and pointing out that the program can still be tested, with a part of it typed and the other not.